From linux-rt-users-owner@vger.kernel.org Fri Aug 24 19:58:47 2007
Return-Path: <linux-rt-users-owner@vger.kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.1.7-deb (2006-10-05) on debian
X-Spam-Level: 
X-Spam-Status: No, score=0.0 required=5.0 tests=AWL autolearn=unavailable 
	version=3.1.7-deb
Received: from vger.kernel.org (vger.kernel.org [209.132.176.167]) by
	mail.tglx.de (Postfix) with ESMTP id A3A6965C292; Fri, 24 Aug 2007 19:58:47
	+0200 (CEST)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand id
	S1762326AbXHXR6Z (ORCPT <rfc822;jan.altenberg@linutronix.de> + 1 other);
	Fri, 24 Aug 2007 13:58:25 -0400
Received: (majordomo@vger.kernel.org) by vger.kernel.org id
	S1765714AbXHXR6Y (ORCPT <rfc822;linux-rt-users-outgoing>); Fri, 24 Aug 2007
	13:58:24 -0400
Received: from ms-smtp-03.nyroc.rr.com ([24.24.2.57]:40924 "EHLO
	ms-smtp-03.nyroc.rr.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org with
	ESMTP id S1765698AbXHXR6U (ORCPT <rfc822;linux-rt-users@vger.kernel.org>);
	Fri, 24 Aug 2007 13:58:20 -0400
Received: from [192.168.23.10] (cpe-24-94-51-176.stny.res.rr.com
	[24.94.51.176]) by ms-smtp-03.nyroc.rr.com (8.13.6/8.13.6) with ESMTP id
	l7OHvEkk027906; Fri, 24 Aug 2007 13:57:17 -0400 (EDT)
Subject: [PATCH RT 2/3] initialize the clock source to jiffies clock.
From:	Steven Rostedt <rostedt@goodmis.org>
To:	Ingo Molnar <mingo@elte.hu>
Cc:	LKML <linux-kernel@vger.kernel.org>, RT <linux-rt-users@vger.kernel.org>, Thomas Gleixner <tglx@linutronix.de>, john stultz <johnstul@us.ibm.com>
Content-Type: text/plain
Date:	Fri, 24 Aug 2007 13:57:14 -0400
Message-Id: <1187978234.2941.18.camel@localhost.localdomain>
Mime-Version: 1.0
X-Mailer: Evolution 2.10.2 
X-Virus-Scanned: Symantec AntiVirus Scan Engine
Sender:	linux-rt-users-owner@vger.kernel.org
Precedence: bulk
X-Mailing-List:	linux-rt-users@vger.kernel.org
X-Filter-To: .Kernel.rt-users
X-Evolution-Source: imap://tglx%40linutronix.de@localhost:8993/
Content-Transfer-Encoding: 8bit

The latency tracer can call clocksource_read very early in bootup and
before the clock source variable has been initialized. This results in a
crash at boot up (even before earlyprintk is initialized). Since the
clock->read variable is points to NULL.

This patch simply initializes the clock to use clocksource_jiffies, so
that any early user of clocksource_read will not crash.

Signed-off-by: Steven Rostedt <rostedt@goodmis.org>

---
 kernel/time/timekeeping.c |    9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

Index: linux-2.6.23.1-rt5/kernel/time/timekeeping.c
===================================================================
--- linux-2.6.23.1-rt5.orig/kernel/time/timekeeping.c
+++ linux-2.6.23.1-rt5/kernel/time/timekeeping.c
@@ -63,8 +63,15 @@ static inline void update_xtime_cache(u6
 #define update_xtime_cache(n) do { } while (0)
 #endif
 
-static struct clocksource *clock; /* pointer to current clocksource */
+extern struct clocksource clocksource_jiffies;
 
+/*
+ * pointer to current clocksource
+ *  Just in case we use clocksource_read before we initialize
+ *  the actual clock source. Instead of calling a NULL read pointer
+ *  we return jiffies.
+ */
+static struct clocksource *clock = &clocksource_jiffies;
 
 #ifdef CONFIG_GENERIC_TIME
 /**
